### PR TITLE
Fix certain overlays and screens blocking home and alt-f4

### LINKED ISSUE #3812: ModSelectOverlay causes the Home button to work incorrectly
**Describe your problem:** The issue basically happens when the `ModSelectOverlay` is open: the `Home` button doesn't send you to the `MainMenu`, but it just closes the `ModSelectOverlay`.

### LINKED ISSUE #9430: Using quick-retry hotkey in results screen when stats are expanded results in black screen
**Describe the bug:** 
Using the quick-retry hotkey in the results screen when the play statistics are expanded results in a black screen, which go backs to the player loader when exited (as seen in the attached video).

Reproduction steps:
1. Play a beatmap and reach the results screen
2. Expand the statistics
3. Press your quick-retry hotkey

**Screenshots or videos showing encountered issue:** 
https://streamable.com/mit0rq
**osu!lazer version:** 
Happens on latest `master`

**Logs:**
Irrelevant here.

### PR DESCRIPTION
**Pull request by Joehuu**
- Resolves https://github.com/ppy/osu/issues/3812
- Fixes https://github.com/ppy/osu/issues/9430, supersedes and closes https://github.com/ppy/osu/pull/9436

### CODE REVIEW

File: osu.Game/Screens/IOsuScreen.cs [hunk 1]
```diff
@@ -56,5 +56,7 @@ public interface IOsuScreen : IScreen
         /// Whether mod rate adjustments are allowed to be applied.
         /// </summary>
         bool AllowRateAdjustments { get; }
+
+        bool OnBackButton();
```
Comment (bdach) [inline osu.Game/Screens/IOsuScreen.cs:60]: needs xmldoc

File: osu.Game/OsuGame.cs [hunk 2]
```diff
@@ -573,7 +573,9 @@ protected override void LoadComplete()
                             Origin = Anchor.BottomLeft,
                             Action = () =>
                             {
-                                if ((ScreenStack.CurrentScreen as IOsuScreen)?.AllowBackButton == true)
+                                var currentScreen = ScreenStack.CurrentScreen as IOsuScreen;
+
+                                if (currentScreen?.AllowBackButton == true && !currentScreen.OnBackButton())
```
Comment (bdach) [inline osu.Game/OsuGame.cs:578]: I'm not sure I'm following why both `AllowBackButton` and `OnBackButton` need to exist. `OnBackButton` seems like the more general mechanism and can probably supersede all usages of `AllowBackButton`, unless I'm missing something.

File: osu.Game/Screens/Multi/Multiplayer.cs [hunk 3]
```diff
@@ -269,6 +263,25 @@ public override bool OnExiting(IScreen next)
...
 
+        public override bool OnBackButton()
+        {
+            var songSelect = screenStack.CurrentScreen as MatchSongSelect;
+
+            if (songSelect?.ModSelect.State.Value == Visibility.Visible)
+            {
+                songSelect.ModSelect.Hide();
+                return true;
+            }
```
Comment (bdach) [inline osu.Game/Screens/Multi/Multiplayer.cs:274]: not sure `Multiplayer` should be prodding its sub-screen states directly like this. maybe something an `OnBackButton()` in `IMultiplayerSubScreen` would be better.

File: osu.Game/Screens/Select/SongSelect.cs [hunk 4]
```diff
@@ -86,7 +86,7 @@ public abstract class SongSelect : OsuScreen, IKeyBindingHandler<GlobalAction>
         [Resolved]
         private BeatmapManager beatmaps { get; set; }
 
-        protected ModSelectOverlay ModSelect { get; private set; }
+        public ModSelectOverlay ModSelect { get; private set; }
```
Comment (bdach) [inline osu.Game/Screens/Select/SongSelect.cs:89]: ideally this shouldn't be done.

**Review by bdach — DISMISSED** [review]

File: osu.Game/OsuGame.cs (hunk 2 above)
Comment (Joehuu) [reply osu.Game/OsuGame.cs:578]: `AllowBackButton` is there to hide the back button and disable its action on certain screens (like player doesn't need a back button, and has local exit logic).

`OnBackButton` is there to override the action, and only logical to implement if `AllowBackButton` is true.

File: osu.Game/Screens/Multi/Multiplayer.cs (hunk 3 above)
Comment (Joehuu) [reply osu.Game/Screens/Multi/Multiplayer.cs:274]: I think it needs to be done here, as sub screens can't control the back button.

```
if ((screenStack.CurrentScreen as IMultiplayerSubScreen).OnBackButton())
	return true;
```

I thought of this, seems more better?

Comment (bdach) [reply osu.Game/Screens/Multi/Multiplayer.cs:274]: yeah, that's what I mean; the check itself being there I don't have much of an issue with, but accessing the subscreens' controls in here (some of which need to be exposed for this to work) doesn't seem too great.

**Comment by smoogipoo** [comment]
I reworded the xmldoc a bit and am fine with the implementation. Only thing I'm a little iffy about is the `OnBackButton()` naming but I can convince myself it makes sense.

Thoughts, @peppy (or merge)

**Review by smoogipoo — APPROVED** [review]

**Comment by peppy** [comment]
The naming matches other similar `OsuScreen` methods so I think it's fine.

**Review by peppy — APPROVED** [review]

**Pull request merged**